lib/internal/map.js

```diff
@@ -1,15 +1,23 @@
 import isArrayLike from 'lodash/isArrayLike';
 import getIterator from './getIterator';
 import noop from 'lodash/noop';
+import okeys from 'lodash/keys';
+import indexOf from 'lodash/_baseIndexOf';
 import once from './once';
 
 export default function _asyncMap(eachfn, arr, iteratee, callback) {
     callback = once(callback || noop);
     arr = arr || [];
-    var results = isArrayLike(arr) || getIterator(arr) ? [] : {};
+    var results = [];
+    var keys;
+    if (!isArrayLike(arr) && !getIterator(arr)) {
+        keys = okeys(arr);
+    }
+
     eachfn(arr, function (value, index, callback) {
         iteratee(value, function (err, v) {
-            results[index] = v;
+            var idx = keys ? indexOf(keys, index, 0) : index;
```

Comment:
this shouldn't be necessary. Just keep track of by creating and incrementing a numeric index variable (`idx` in the eachfn

```js
var idx = 0;
eachfn(arr, function(...) {
   idx += 1;
   iteratee(...
```